tests/src/Interop/ArrayMarshalling/SafeArray/SafeArrayNative.cpp

```diff
@@ -0,0 +1,651 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include <xplatform.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include "platformdefines.h"
+
+//--------------------------------
+//	Start of SafeArray calls
+//--------------------------------
+
+extern "C" DLL_EXPORT BOOL STDMETHODCALLTYPE SafeArray_In(SAFEARRAY* psa)
+{
+	long i, lUbound, lLbound;
+	int  *pInt; //will point to the data
+	HRESULT hr;
+
+	// Get a pointer to the elements of the array.
+	hr = SafeArrayAccessData(psa, (void **)&pInt);
+	if(FAILED(hr))
+	{
+		printf("\t\tSafeArrayAccessData call in SafeArray_In failed!\n");
+		return FALSE;
+	}
+
+	//get upperbound
+	hr = SafeArrayGetUBound(psa, 1, &lUbound);
+	if(FAILED(hr))
+	{
+		printf("\t\tSafeArrayGetUBound call in SafeArray_In failed!\n");
+		return FALSE;
+	}
+
+	//check upperbound
+	if(lUbound != 255) //since num of elems = lUBound - lLBound + 1
+	{
+		printf("\t\tlUbound not as expected in SafeArray_In!\n");
+		printf("\t\t\tlUbound = %d",lUbound);
+		return FALSE;
+	}
+
+	//get lowerbound
+	hr = SafeArrayGetLBound(psa, 1, &lLbound);
+	if(FAILED(hr))
+	{
+		printf("\t\tSafeArrayGetLBound call in SafeArray_In failed!\n");
+		return FALSE;
+	}
+
+	//check lowerbound
+	if(lLbound != 0) 
+	{
+		printf("\t\tlLbound not as expected in SafeArray_In!\n");
+		printf("\t\t\tlLbound = %d",lLbound);
+		return FALSE;
+	}
+
+	//check dimension
+	if(SafeArrayGetDim(psa) != 1)
+	{
+		printf("\t\tDimension not as expected in SafeArray_In!\n");
+		return FALSE;
+	}
+
+	//check element size
+	if(SafeArrayGetElemsize(psa) != 4) //size of each element should be 4 bytes
```

Comment:
Why? This is a perfect example of why I dislike how these tests are written. The base invariant here is that the passed in SafeArray is of INT4, but that isn't defined anywhere in this code nor validated. Additionally this could be for a float for that matter. Basically I am fine with this kind of validation, but my question to you and something we should always consider when writing tests is the following list:

What are we validating?
Why are we validating it?
If this test fails what does it indicate?
If this test fails is it obvious what other scenarios are impacted?
Is this test validating undocumented or unsupported scenarios?
Can a new person look at this test and easily determine the test purpose and value?